ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -164,12 +206,38 @@ def euclidean(cls, pos1, pos2):
     @classmethod
     def distance(cls, waypoints, wp1, wp2):
         dl = 0
-        for i in range(wp1, wp2+1):
-            dl += cls.euclidean(waypoints[wp1].pose.pose.position,
-                                waypoints[i].pose.pose.position)
+        n = len(waypoints)
+        for i in range(wp1+1, wp2):
+            dl += cls.euclidean(waypoints[wp1 % n].pose.pose.position,
+                                waypoints[i % n].pose.pose.position)
             wp1 = i
         return dl
 
+    def dist_to_tl(self):
+        diff = self.tl_idx - self.next_idx
+        if diff < 0:
+            diff += self.n_waypoints
+        return diff
+
+    def decelerate(self, waypoints):
+        dist = self.dist_to_tl()
+        if self.tl_idx > - 1 and dist < LOOKAHEAD_WPS:
+            prev_wp = waypoints[dist]
+            self.set_waypoint_velocity(waypoints, dist, 0.)
+            for i in range(dist-1, -1, -1):
+                wp = waypoints[i]
+                prev_vel = self.get_waypoint_velocity(prev_wp)
+                d = self.distance(waypoints, i, i + 1)
+                # Approximate time between two waypoints
+                t = max(d / (prev_vel + MAX_DECEL), .1)
+                old_vel = self.get_waypoint_velocity(wp)
+                vel = prev_vel + MAX_DECEL * t
+                vel = vel if vel > .2 else .0
```

Comment:
Maybe create a variable for this 0.2? A bit easier to read and keep track of parameters.

Reply:
This one is more like a constant which need not be controlled. I will create a constant similar to `MAX_DECEL` (also the latter one will be removed in favour of reading a parameter).